--- lbreakout2-2.3.2/client/Makefile.am.orig	Tue Aug  6 21:00:02 2002
+++ lbreakout2-2.3.2/client/Makefile.am	Wed Sep  4 22:54:41 2002
@@ -44,7 +44,7 @@
 credit.c audio.c audio.h value.h item.h menu.h manager.h value.c item.c menu.c \
 manager.c chart.h chart.c editor.h editor.c help.h help.c  \
 hint.c hint.h theme.c theme.h client.h gui.h windows.h image.h \
-display.h comm.h png.h \
+display.h comm.h pngm.h \
 empty.hscr
 
 SUBDIRS = gfx sounds levels
--- lbreakout2-2.3.2/client/sdl.c.orig	Tue Aug 13 17:32:56 2002
+++ lbreakout2-2.3.2/client/sdl.c	Wed Sep  4 22:54:52 2002
@@ -19,7 +19,7 @@
 #include <stdlib.h>
 #include <string.h>
 #include "sdl.h"
-#include "png.h"
+#include "pngm.h"
 
 extern int term_game;
 
